[1.15.x] Added event for when a particle is spawned in the world

ChristianANoeddeboe opened the pull request:
Adds a new event, which is used when a particle is spawned in the world.

This allows players to temporarily replace or change a spawned vanilla particle.

An example scenario is increasing the damage the player does. So far a high damage hit on an entity will display many damage particles, and this will help modders reduce these to a reasonable amount.

Testmod has been added as ParticleWorldHandleEventTest.java.

Commit:
Added event for when a particle is spawned in the world

GotoLink commented:
The event javadoc is talking about command, probably an error ?
The second argument is always false according to the patches. What happens if you force longDistance to true by using this event ?

ChristianANoeddeboe committed:
Update ParticleWorldHandleEvent.java

Changed the javadocs to not reference CommandEvent.

ChristianANoeddeboe commented:
> 
> 
> The event javadoc is talking about command, probably an error ?
> The second argument is always false according to the patches. What happens if you force longDistance to true by using this event ?

Changing longDistance will just increase the range in which players can see the particle. Will not break anything, and may be useful, if you want to give the particle a bigger "effect".

marchermans commented on patches/minecraft/net/minecraft/world/server/ServerWorld.java.patch:
```diff
@@ -306,7 +306,37 @@
+ 
+    public <T extends IParticleData> int func_195598_a(T p_195598_1_, double p_195598_2_, double p_195598_4_, double p_195598_6_, int p_195598_8_, double p_195598_9_, double p_195598_11_, double p_195598_13_, double p_195598_15_) {
+-      SSpawnParticlePacket sspawnparticlepacket = new SSpawnParticlePacket(p_195598_1_, false, p_195598_2_, p_195598_4_, p_195598_6_, (float)p_195598_9_, (float)p_195598_11_, (float)p_195598_13_, (float)p_195598_15_, p_195598_8_);
++      net.minecraftforge.event.ParticleWorldHandleEvent event = new net.minecraftforge.event.ParticleWorldHandleEvent(p_195598_1_, false, p_195598_8_, p_195598_2_, p_195598_4_, p_195598_6_, p_195598_9_, p_195598_11_, p_195598_13_, p_195598_15_);
```
Move the construction of the event, and the posting to the event bus to the `ForgeEventFactory`. This keeps all the events in one place and reduces the patch size here from 3 to 2, every little bit helps.

marchermans commented on patches/minecraft/net/minecraft/world/server/ServerWorld.java.patch:
```diff
@@ -306,7 +306,37 @@
+    public <T extends IParticleData> boolean func_195600_a(ServerPlayerEntity p_195600_1_, T p_195600_2_, boolean p_195600_3_, double p_195600_4_, double p_195600_6_, double p_195600_8_, int p_195600_10_, double p_195600_11_, double p_195600_13_, double p_195600_15_, double p_195600_17_) {
+-      IPacket<?> ipacket = new SSpawnParticlePacket(p_195600_2_, p_195600_3_, p_195600_4_, p_195600_6_, p_195600_8_, (float)p_195600_11_, (float)p_195600_13_, (float)p_195600_15_, (float)p_195600_17_, p_195600_10_);
+-      return this.func_195601_a(p_195600_1_, p_195600_3_, p_195600_4_, p_195600_6_, p_195600_8_, ipacket);
++      net.minecraftforge.event.ParticleWorldHandleEvent event = new net.minecraftforge.event.ParticleWorldHandleEvent(p_195600_2_, false, p_195600_10_, p_195600_4_, p_195600_6_, p_195600_8_, p_195600_11_, p_195600_13_, p_195600_15_, p_195600_17_);
```
Again this can probably be handled better by using a method in `ForgeEventFactory` possibly returning an Optional which is empty when the event is cancelled, or contains the even if it is allowed to pass.

Would again reduce the patch size.

ChristianANoeddeboe replied:
EDIT: I've incorporated the changes, please tell me again if I need to change anything 😃 

I'm having a bit of trouble understanding how the optional will reduce patch size. Using an optional I get this in the ServerWorld.java:
```
public <T extends IParticleData> int spawnParticle(T type, double posX, double posY, double posZ, int particleCount, double xOffset, double yOffset, double zOffset, double speed) {
      AtomicInteger i = new AtomicInteger();
      ForgeEventFactory.onWorldParticleSpawn(type, false, posX, posY, posZ, particleCount, xOffset, yOffset, zOffset, speed).ifPresent(event -> {
         SSpawnParticlePacket sspawnparticlepacket = new SSpawnParticlePacket(event.getParticle(), event.isLongDistance(), (float)event.getXCoord(), (float)event.getYCoord(), (float)event.getZCoord(), (float)event.getXOffset(), (float)event.getYOffset(), (float)event.getZOffset(), (float)event.getParticleSpeed(), event.getParticleCount());
         for (ServerPlayerEntity serverplayerentity : this.players) {
            if (this.sendPacketWithinDistance(serverplayerentity, event.isLongDistance(), event.getXCoord(), event.getYCoord(), event.getZCoord(), sspawnparticlepacket)) {
               i.incrementAndGet();
            }
         }
      });
      return i.get();
   }

   public <T extends IParticleData> boolean spawnParticle(ServerPlayerEntity player, T type, boolean longDistance, double posX, double posY, double posZ, int particleCount, double xOffset, double yOffset, double zOffset, double speed) {
      AtomicBoolean result = new AtomicBoolean(false);
      ForgeEventFactory.onWorldParticleSpawn(type, longDistance, posX, posY, posZ, particleCount, xOffset, yOffset, zOffset, speed).ifPresent(event -> {
         SSpawnParticlePacket sspawnparticlepacket = new SSpawnParticlePacket(event.getParticle(), event.isLongDistance(), (float)event.getXCoord(), (float)event.getYCoord(), (float)event.getZCoord(), (float)event.getXOffset(), (float)event.getYOffset(), (float)event.getZOffset(), (float)event.getParticleSpeed(), event.getParticleCount());
         result.set(this.sendPacketWithinDistance(player, event.isLongDistance(), event.getXCoord(), event.getYCoord(), event.getZCoord(), sspawnparticlepacket));
      });
      return result.get();
   }
```

With the following in the ForgeEventFactory.java:
```
public static Optional<ParticleWorldHandleEvent> onWorldParticleSpawn(IParticleData particle, boolean longDistance, double posX, double posY, double posZ, int particleCount, double xOffset, double yOffset, double zOffset, double speed) {
        net.minecraftforge.event.ParticleWorldHandleEvent event = new net.minecraftforge.event.ParticleWorldHandleEvent(particle, longDistance, particleCount, speed, posX, posY, posZ, xOffset, yOffset, zOffset);
        if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event)) return Optional.empty();
        return Optional.of(event);
    }
```

marchermans reviewed (changes requested):
You can reduce your patch sizes if you use create a relevant method in `ForgeEventFactory` and process it a bit there. Then you would only need to check the result of that method (which could be an optional, containing the fired event if it is allowed to pass) in the patched code. Reducing it from a multiline patch to a single line patch.

marchermans reviewed (changes requested):
Also a testmod is required, since this adds an event and modders should have an example on how to use it,

Commit:
Created test mod showing an example usecase.
Added method to ForgeEventFactory.java, which creates and posts the event, returning an optional containing either the event, or empty if the event has been cancelled.
Rewrote handling of the event in ServerWorld.java.

Commit:
Merge branch '1.15.x_ParticleEvent' of https://github.com/ChristianANoeddeboe/MinecraftForge into 1.15.x_ParticleEvent

Commit:
Forgot to add license to the new files.

diesieben07 commented on patches/minecraft/net/minecraft/world/server/ServerWorld.java.patch:
```diff
@@ -306,7 +306,45 @@
+-
+-      return i;
++      });
++      return i.get();
```
You can make this patch _way_ smaller by not using `Optional`. Make `onParticleSpawn` return the event, or null. Then just have one if statement that returns immediately if the event is cancelled (two lines so far). After that just re-assign the parameter variables using the event. Then your patch becomes only additions, which is much easier to maintain.

marchermans replied:
@ChristianANoeddeboe sorry to indeed have send you on a goose chase, i misread the PR and the patch. Using an Optional here is indeed not usefull. But you will need to have the method in ForgeEventFactory @diesieben07 is completely right here.

The best approach would probably be to combine what you had in the first place with the factory method either returning null to cancel the behaviour, or returning the processed event. This would indeed make it a two, possibly three, line patch.

Sorry for the inconvenience.

diesieben07 commented on patches/minecraft/net/minecraft/world/server/ServerWorld.java.patch:
```diff
@@ -306,7 +306,45 @@
++         SSpawnParticlePacket sspawnparticlepacket = new SSpawnParticlePacket(event.getParticle(), event.isLongDistance(), (float)event.getXCoord(), (float)event.getYCoord(), (float)event.getZCoord(), (float)event.getXOffset(), (float)event.getYOffset(), (float)event.getZOffset(), (float)event.getParticleSpeed(), event.getParticleCount());
++         result.set(this.func_195601_a(p_195600_1_, event.isLongDistance(), event.getXCoord(), event.getYCoord(), event.getZCoord(), sspawnparticlepacket));
++      });
++      return result.get();
```
See above.

diesieben07 commented on src/main/java/net/minecraftforge/event/ForgeEventFactory.java:
```diff
@@ -528,6 +529,12 @@ public static void onPlayerBrewedPotion(PlayerEntity player, ItemStack stack)
     }
 
+    public static Optional<ParticleWorldHandleEvent> onWorldParticleSpawn(IParticleData particle, boolean longDistance, double posX, double posY, double posZ, int particleCount, double xOffset, double yOffset, double zOffset, double speed) {
+        net.minecraftforge.event.ParticleWorldHandleEvent event = new net.minecraftforge.event.ParticleWorldHandleEvent(particle, longDistance, particleCount, speed, posX, posY, posZ, xOffset, yOffset, zOffset);
```
This is not a patch file, no need to use FQNs.

diesieben07 commented on src/main/java/net/minecraftforge/event/ParticleWorldHandleEvent.java:
```diff
@@ -0,0 +1,142 @@
+ * This event is fired on the {@link MinecraftForge#EVENT_BUS}.<br>
+ **/
+@Cancelable
+public class ParticleWorldHandleEvent extends Event {
```
This is a bad name, imho. You should make it clear that:
1. This is a server-side event.
2. That this is about particle spawns.

My suggestion would be `ServerParticleSpawnEvent`.

marchermans replied:
Probably good idea too, I initially thought it was fired on both sides, but I checked the code myself this morning and yeah a name indicating that this is a server side only event and what it is fired for, would be great.

Commit:
Changed event name.
Removed optional from EventFactory.

Commit:
Changed event name.
Removed optional from EventFactory.

Commit:
Changed event name.
Removed optional from EventFactory.
Changed ServerWorld.java to use the changes above.

ChristianANoeddeboe commented:
Thank you for your help @diesieben07 and @OrionDevelopment, I've changed the code now based on your feedback.

diesieben07 commented on src/main/java/net/minecraftforge/event/ForgeEventFactory.java:
```diff
@@ -528,6 +529,12 @@ public static void onPlayerBrewedPotion(PlayerEntity player, ItemStack stack)
 
+    public static ServerParticleSpawnEvent onWorldParticleSpawn(IParticleData particle, boolean longDistance, double posX, double posY, double posZ, int particleCount, double xOffset, double yOffset, double zOffset, double speed) {
+        ServerParticleSpawnEvent event = new ServerParticleSpawnEvent(particle, longDistance, particleCount, speed, posX, posY, posZ, xOffset, yOffset, zOffset);
+        if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event)) return null;
```
This is not a patch file. You don't need to use fully qualified names.

diesieben07 commented on src/main/java/net/minecraftforge/event/ServerParticleSpawnEvent.java:
```diff
@@ -0,0 +1,142 @@
+import net.minecraftforge.eventbus.api.Event;
+
+/**
+ * ParticleWorldHandleEvent is fired before an {@link SSpawnParticlePacket} is sent.<br>
```
Missing name change here.

diesieben07 commented on src/main/java/net/minecraftforge/event/ServerParticleSpawnEvent.java:
```diff
@@ -0,0 +1,142 @@
+ * ParticleWorldHandleEvent is fired before an {@link SSpawnParticlePacket} is sent.<br>
+ * <br>
+ * This event is {@link Cancelable}. <br>
+ * If the event is canceled, the execution of the command does not occur.<br>
```
Copy-paste error?

ChristianANoeddeboe replied:
Yep sorry about that, I've changed it now.

diesieben07 reviewed (changes requested):
Few small changes needed still, other than that looks good now.

Commit:
Changed wrong name in comments and removed fully qualified name.

diesieben07 commented on src/test/java/net/minecraftforge/debug/world/ServerParticleSpawnEventTest.java:
```diff
@@ -0,0 +1,46 @@
+     */
+    @SubscribeEvent
+    public static void onEvent(ServerParticleSpawnEvent event) {
+        if(event.getParticle().getParameters().equals(ParticleTypes.DAMAGE_INDICATOR.getParameters())) {
```
missing space after if

ChristianANoeddeboe replied:
Yeah, sorry I completely forgot to check what formatting you already used... I think I got them all corrected now.

diesieben07 reviewed (changes requested):
Sorry, another thing I missed:
all your files are not formatted correctly, braces need to be on new lines.
please check the other files in Forge.

Commit:
Changed formatting to match the standard used.

diesieben07 commented:
1.15 is the current LTS version, please port your PR to 1.16. Features go in the current version first before being backported.

ChristianANoeddeboe commented:
I've created a 1.16 version of this now. @diesieben07

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=6805) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=6805) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: ChristianANoeddeboe<br/>:x: Christian<br/><hr/>**Christian** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MinecraftForge/MinecraftForge?pullRequest=6805) it.</sub>
